packages/ui/src/components/va-virtual-scroller/VaVirtualScroller.vue

```diff
@@ -0,0 +1,126 @@
+<template>
+  <div ref="wrapper" class="va-virtual-scroller">
+    <div class="va-virtual-scroller__container" :style="containerStyleComputed">
+      <div
+        ref="list"
+        role="list"
+        class="va-virtual-scroller__list"
+        :class="listClassComputed"
+        :style="listStyleComputed"
+      >
+        <template
+          v-for="(item, index) in renderBuffer"
+          :key="uniqueKey(item, index)"
+        >
+            <slot v-bind="{item, index: renderStartIndex + index}" />
+        </template>
+      </div>
+    </div>
+  </div>
+</template>
+
+<script lang="ts">
+import { defineComponent, ref, computed } from 'vue'
+import pick from 'lodash/pick.js'
+import isArray from 'lodash/isArray.js'
+
+import { useEvent, useBem } from '../../composables'
+import { useVirtualScrollerSizes, useVirtualScrollerSizesProps } from './useVirtualScrollerSizes'
+
+import { warn } from '../../services/utils'
+
+export default defineComponent({
+  name: 'VaVirtualScroller',
+
+  props: {
+    ...useVirtualScrollerSizesProps,
+    items: { type: Array, required: true, validator: (v: any[]) => !!v.length },
+    bench: { type: Number, default: 10, validator: (v: number) => v >= 0 },
+    trackBy: { type: String, default: '' },
+    disabled: { type: Boolean, default: false },
+  },
+
+  setup: (props) => {
+    const currentListScroll = ref<number>(0)
+    const handleScroll = () => {
+      if (!wrapper.value) { return }
+      currentListScroll.value = wrapper.value[props.horizontal ? 'scrollLeft' : 'scrollTop']
+    }
+    if (!props.disabled) { useEvent('scroll', handleScroll, true) }
+
+    const { list, wrapper, wrapperSize, itemSize } = useVirtualScrollerSizes(props, currentListScroll)
+
+    const uniqueKey = (item: unknown, index: number) => {
+      if (props.trackBy && item && typeof item === 'object' && !isArray(item)) {
+        const key = (item as Record<string, unknown>)[props.trackBy]
+
+        if (key || key === 0) { return `virtual-scroller-item-${key}` }
```

Comment:
Probably doesn't matter, but maybe shorter key? like `va-${key}`. Btw, why did you even use string for that? key could be an object.

Reply:
> Key should be unique. There is may be another `va-1` key on page.

Actually it should be unique only for current list, not on page. I don't see a way for you can have same items by key in list. I actually even don't see why prefix is needed at all.

> object key

Checked docs - https://vuejs.org/api/built-in-special-attributes.html.
Seems you're right, can't be object :D (only string, number or symbol). Still don't see a need to convert or validate it - vue can do that instead.

I used objects as keys before, but that is kinda undocumented feature. :)